spring-cloud-gcp-autoconfigure/src/test/java/org/springframework/cloud/gcp/autoconfigure/pubsub/it/PubSubChannelAdaptersIntegrationTests.java

```diff
@@ -201,6 +205,57 @@ public void sendAndReceiveMessageManualAck() {
 		});
 	}
 
+	// If this test flakes, delete it.
+	// It verifies that in AUTO_ACK mode, the message is neither acked nor nacked, and that
+	// redelivery happens after subscription's ackDeadline passes.
+	// There is also a client library bug (https://github.com/googleapis/java-pubsub/issues/141) that
+	// results in ackDeadline being extended by 60 seconds even when maxAckExtensionPeriod is zero,
+	// making minimum redelivery time is ackDeadline + 60.
+	@Test
+	public void sendAndReceiveMessageAutoAckWithFailure() {
+		ApplicationContextRunner customContextRunner = new ApplicationContextRunner()
+				.withConfiguration(AutoConfigurations.of(
+						GcpContextAutoConfiguration.class,
+						GcpPubSubAutoConfiguration.class))
+				.withUserConfiguration(SubscribableConfiguration.class, CommonConfiguration.class);
+
+		customContextRunner.run((context) -> {
+			try {
+				context.getBean(PubSubInboundChannelAdapter.class).setAckMode(AckMode.AUTO_ACK);
+				context.getBean("inputChannel", MessageChannel.class).send(
+						MessageBuilder.withPayload("This message is in trouble.".getBytes()).build());
+
+				SubscribableChannel channel = context.getBean("outputChannel", SubscribableChannel.class);
+
+				AtomicInteger numReceivedMessages = new AtomicInteger(0);
+				channel.subscribe(msg -> {
+					if (numReceivedMessages.incrementAndGet() == 1) {
+						throw new RuntimeException("BOOM!");
+					}
+				});
+
+				// wait for initial delivery
+				Awaitility.await().atMost(10, TimeUnit.SECONDS)
+						.until(() -> numReceivedMessages.get() > 0);
+				assertThat(numReceivedMessages.get()).isEqualTo(1);
+
+				// Expect redelivery within 2 minutes:
+				// 30 seconds subscription ackDeadline
+				// + 60 seconds https://github.com/googleapis/java-pubsub/issues/141
+				// + 30 seconds anti-flake buffer
+				Awaitility.await().atMost(120, TimeUnit.SECONDS)
```

Comment:
>It verifies that in AUTO_ACK mode, the message is neither acked nor nacked
How does it verify that it's not nacked?

If this test in fact takes around 2 minutes to pass, I think it's too much to add to the build.
Can we just verify that the message is not redelivered immediately because that was the original problem anyway?
In fact, it doesn't look like this test is verifying that aspect at all? The message can be redelivered immediately, and it will pass.

Reply:
In theory, `assertThat(numReceivedMessages.get()).isEqualTo(1);` could catch immediate redelivery, but I don't think it would be reliable. 

I had another check there waiting for 10 seconds and making sure the message count was still 1, but that relied on Thread.sleep(). I think invariants verfiication was added into a later version of Awaitility... going to try using that.

Reply:
I'd like to verify eventual redelivery (provided I can do it without creating a flake) since we also had a bug where the message got accidentally acked, and so never redelivered.

Reply:
Actually, await().atLeast() does the job.